### PR TITLE
add meta client 2.0

### PR DESCRIPTION
**Pull request by Nicole00**

### CODE REVIEW

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+     * get all parts of one space
+     *
+     * @param spaceName nebula graph space name
+     * @return List
+     */
+    public List<Integer> getSpaceParts(String spaceName) {
+        if (!metaInfo.getSpacePartLocation().containsKey(spaceName)) {
+            fillMetaInfo();
+        }
+        Set<Integer> spaceParts = metaInfo.getSpacePartLocation().get(spaceName).keySet();
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:346]: Parttion will not be repeated

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+
+    /**
+     * fresh edges in meta info
+     */
+    private void freshEdges(String spaceName) {
+        List<EdgeItem> edgeItems = null;
+        try {
+            edgeItems = metaClient.getEdges(spaceName);
+        } catch (TException | ExecuteFailedException e) {
+            LOGGER.error("fresh edge error, ", e);
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:459]: add return

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+
+    /**
+     * fresh tags in meta info
+     */
+    private void freshTags(String spaceName) {
+        List<TagItem> tagItems = null;
+        try {
+            tagItems = metaClient.getTags(spaceName);
+        } catch (TException | ExecuteFailedException e) {
+            LOGGER.error("fresh tag error, ", e);
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:415]: add return

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+
+    /**
+     * get all parts of one space
+     *
+     * @param spaceName nebula graph space name
+     * @return List
+     */
+    public List<Integer> getSpaceParts(String spaceName) {
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:311]: List\<int\>

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+
+    private MetaInfo metaInfo = new MetaInfo();
+
+    private MetaManager() {
+    }
+
+    /**
+     * only way to get a MetaManager object
+     */
+    public static MetaManager getMetaManager(List<HostAndPort> address) throws TException {
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:48]: unify use `com.vesoft.nebula.client.graph.data.HostAddress`, then the users only import our supported modules.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+
+    /**
+     * get edge id
+     *
+     * @param spaceName nebula graph space name
+     * @param edgeName  nebula edge name
+     * @return long
+     */
+    private long getEdgeId(String spaceName, String edgeName) {
+        if (!metaInfo.getTagNameMap().containsKey(spaceName)
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: the `read` also needs to add a lock. you can use the lock.readLock()

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+ * How to use:
+ * MetaManager manager = MetaManager.getMetaManager(Arrays.asList(HostAndPort.fromString(addr)));
+ * manager.init();
+ */
+public class MetaManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MetaManager.class);
+
+    private static MetaClient metaClient;
+    private static MetaManager metaManager;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:37]: why not move the lock to MetaInfo, the MetaInfo make itself thread-safe. and the MetaManager only call the `get` or `set` function by MetaInfo

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 4 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:311]: it's illegal java List

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 6 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: read does not need to lock, only write cause map's endless loop

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 5 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:48]: ok

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 7 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:37]: no need,  the set functions are just put operators.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 6 above)
Comment (laura-ding) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: the map's value is the list, the list will change, such as the schema's columns. it means you read the list and rewrite the list.

Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: if It happens，it means the map‘s value need to be update.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 4 above)
Comment (laura-ding) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:310]: `int[]`

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 6 above)
Comment (laura-ding) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: ok,  Java's garbage collection mechanism makes it safe. But it's still not safe. if the key not exist, here will multi-thread call fillMetaInfo(), and the fillMetaInfo() is not thread-safe.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+        }
+        return metaManager;
+    }
+
+    private void getClient(List<HostAddress> address) throws TException {
+        if (metaClient == null) {
+            synchronized (this) {
+                if (metaClient == null) {
+                    metaClient = new MetaClient(address);
+                    metaClient.connect();
```
Comment (laura-ding) [inline client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:65]: You can move tow line to line 51, this function is no need.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 6 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:168]: fillMetaInfo() is thread safe, this method is synchronized.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 8 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:65]: in order to get exactly one metaClient, i used double check for metaClient. Split the `new` operations of metaManager and metaClient may be more clear.

File: client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java (hunk 4 above)
Comment (Nicole00) [reply client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java:310]: no need to convert set to array.

**Pull request merged**